## BUSA8001- Programming Task 1  

**Assignment Points**: 100  
**Submission**: Provide your answers in this notebook and submit it via iLearn

- Where a question requires a written answer provide your solution in Markdown in the cells under each question.
- Comment out your print statements unless you are explicitly asked to print your output. 
- 5 marks will be deducted for printed outputs that are not asked for.

### About the Assignment

- For this assignment there are two files in the `data` folder `credit_record.csv` and `application_record.csv` where bank clients are related by the `ID` column.

- In `application_record.csv` we have the following variables

| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| AMT_INCOME   | Annual income  |  |
| NAME_INCOME_TYPE   | Income Source |  |
| NAME_EDUCATION_TYPE   | Level of Education  |  |
| CODE_GENDER   | Applicant's Gender   |  |
| FLAG_OWN_CAR | Car Ownership |  | 
| CNT_CHILDREN | Number of Children | |
| FLAG_OWN_REALTY | Real Estate Ownership | | 
| NAME_FAMILY_STATUS | Relationship Status | | 
| NAME_HOUSING_TYPE | Housing Type | | 
| DAYS_BIRTH | No. of Days | Count backwards from current day (0), -1 means yesterday
| DAYS_EMPLOYED | No. of Days | Count backwards from current day (0). If positive, it means the person is currently unemployed.
| FLAG_MOBIL | Mobile Phone Ownership | | 
| FLAG_WORK_PHONE | Work Phone Ownership | | 
| FLAG_PHONE | Landline Phone Ownership | | 
| FLAG_EMAIL | Landline Phone Ownership | | 
| OCCUPATION_TYPE | Occupation | | 
| CNT_FAM_MEMBERS | Count of Family Members | |



- In `credit_record.csv` we have the following variables


| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number | |
| MONTHS_BALANCE | Number of months in the past from now when STATUS is measured | 0 = current month, -1 = last month, -2 = two months ago, etc.|
| STATUS | Number of days a payment is past due | 0: 1-29 days past due 1: 30-59 days past due 2: 60-89 days overdue 3: 90-119 days overdue 4: 120-149 days overdue 5: Overdue or bad debts, write-offs for more than 150 days C: paid off that month X: No loan for the month |

---
---

### Task 1: Reading, Summarising and Cleaning Data (Total Marks: 30)

**Question 1.** 

1. Import the `application_record.csv` and `credit_record.csv` files from `data` folder into pandas DataFrames named `df_application` and `df_credit`, respectively. (1 mark)

2. How many rows are there in `df_application` and `df_credit`, respectively? Provide your answers with print() and state them in Markdown text. (1 mark)

3. How many unique bank clients are there in `df_application` and `df_credit`? Provide your answers with print() and state them in Markdown text. (1 mark)

4. Add the records from `df_credit` to `df_application` by merging the data from the two DataFrames on the `ID` column, and output the joint data into a new DataFrame named `df`. Hint: Use `merge` function from pandas by setting `how` parameter to `inner` (4 marks) 

5. How many rows and how many unique clients are there now in `df`? (1 mark)

6. How are multiple rows for each `ID` in `df` different? Answer in Markdown text. (2 mark) 

(10 marks)

In [1]:
# ---- provide your code here -----
# Question 1: 
# 1.Import the application_record.csv and credit_record.csv files from data folder into pandas DataFrames 
# named df_application and df_credit, respectively.

import pandas as pd
import numpy as np

# Load the CSV files
df_application = pd.read_csv('data/application_record.csv')
df_credit = pd.read_csv('data/credit_record.csv')

#df_application.head(10)
#df_credit.head(10)

In [2]:
# 2. How many rows are there in df_application and df_credit, respectively? 
# Provide your answers with print() and state them in Markdown text.

print("Number of rows in df_application:", df_application.shape[0])
print("Number of rows in df_credit:", df_credit.shape[0])


Number of rows in df_application: 438445
Number of rows in df_credit: 1047185


---
---
2.

Number of rows in df_application: 438445

Number of rows in df_credit:1047185

In [3]:
# 3. How many unique bank clients are there in df_application and df_credit? 
# Provide your answers with print() and state them in Markdown text.

print("Number of unique bank clients in df_application:", df_application['ID'].nunique())
print("Number of unique bank clients in df_credit:", df_credit['ID'].nunique())

Number of unique bank clients in df_application: 438398
Number of unique bank clients in df_credit: 45924


---
---
3.

Number of unique bank clients in df_application: 438398

Number of unique bank clients in df_credit:45924

In [4]:
# 4. Add the records from df_credit to df_application by merging the data from the two DataFrames on the ID column, 
# and output the joint data into a new DataFrame named df 
# Hint: Use merge function from pandas by setting how parameter to inner (4 marks)

df = pd.merge(df_application, df_credit, on='ID', how='inner')

df.head(25)

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0   5008804           M            Y               Y           0.0   
1   5008804           M            Y               Y           0.0   
2   5008804           M            Y               Y           0.0   
3   5008804           M            Y               Y           0.0   
4   5008804           M            Y               Y           0.0   
5   5008804           M            Y               Y           0.0   
6   5008804           M            Y               Y           0.0   
7   5008804           M            Y               Y           0.0   
8   5008804           M            Y               Y           0.0   
9   5008804           M            Y               Y           0.0   
10  5008804           M            Y               Y           0.0   
11  5008804           M            Y               Y           0.0   
12  5008804           M            Y               Y           0.0   
13  5008804           M            Y               Y           0.0   
14  5008804           M            Y               Y           0.0   
15  5008804           M            Y               Y           0.0   
16  5008805           M            Y               Y           0.0   
17  5008805           M            Y               Y           0.0   
18  5008805           M            Y               Y           0.0   
19  5008805           M            Y               Y           0.0   
20  5008805           M            Y               Y           0.0   
21  5008805           M            Y               Y           0.0   
22  5008805           M            Y               Y           0.0   
23  5008805           M            Y               Y           0.0   
24  5008805           M            Y               Y           0.0   

    AMT_INCOME NAME_INCOME_TYPE NAME_EDUCATION_TYPE NAME_FAMILY_STATUS  \
0     427500.0          Working    Higher education     Civil marriage   
1     427500.0          Working    Higher education     Civil marriage   
2     427500.0          Working    Higher education     Civil marriage   
3     427500.0          Working    Higher education     Civil marriage   
4     427500.0          Working    Higher education     Civil marriage   
5     427500.0          Working    Higher education     Civil marriage   
6     427500.0          Working    Higher education     Civil marriage   
7     427500.0          Working    Higher education     Civil marriage   
8     427500.0          Working    Higher education     Civil marriage   
9     427500.0          Working    Higher education     Civil marriage   
10    427500.0          Working    Higher education     Civil marriage   
11    427500.0          Working    Higher education     Civil marriage   
12    427500.0          Working    Higher education     Civil marriage   
13    427500.0          Working    Higher education     Civil marriage   
14    427500.0          Working    Higher education     Civil marriage   
15    427500.0          Working    Higher education     Civil marriage   
16    427500.0          Working    Higher education     Civil marriage   
17    427500.0          Working    Higher education     Civil marriage   
18    427500.0          Working    Higher education     Civil marriage   
19    427500.0          Working    Higher education     Civil marriage   
20    427500.0          Working    Higher education     Civil marriage   
21    427500.0          Working    Higher education     Civil marriage   
22    427500.0          Working    Higher education     Civil marriage   
23    427500.0          Working    Higher education     Civil marriage   
24    427500.0          Working    Higher education     Civil marriage   

   NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  FLAG_WORK_PHONE  \
0   Rented apartment      -12005          -4542           1                1   
1   Rented apartment      -12005          -4542           1                1   
2   Rented apartment   

In [5]:
# 5. How many rows and how many unique clients are there now in df?

print("Number of rows in df:", df.shape[0])
print("Number of unique clients in df:", df['ID'].nunique())

Number of rows in df: 776325
Number of unique clients in df: 36396


6. How are multiple rows for each ID in df different? Answer in Markdown text.

Although these IDs have same information like gender, income type, education type, etc., some IDs have different values in  "MONTH_BALANCE" AND 'STATUS' columns.

---- provide your written answer here ----


---

**Question 2.**

1. Change the values of `STATUS` in `df` according to the following mapping: {C, X, 0} -> 0 and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers. (2 marks)

2. Create a new *numpy* array called `list_of_past_due` that includes the unique ID numbers of clients whose `STATUS = 1` at any point during the last 12 months (hint: count the current month as the first month). (2 marks) 

3. Create a new DataFrame called `df_final` that contains the rows of `df` for which the `ID` is in `list_of_past_due`, keeping only one row for each `ID` (hint: keep the first duplicate row). How many rows do you have in `df_final`? Answer using both print() function and in Markdown text. (hint: find out about `isin()` function in pandas.) (2 marks)

4. Add a new column `y = 1` for all the rows in `df_final`. (1 marks)

5. Increase `df_final` to a total of 4,500 rows by adding rows from `df` with unique `ID`s which are not in `list_of_past_due`. To do this start adding the rows from the beginning of `df`. (hint: learn what `~`, i.e. tilde sign, does in pandas). (2 marks) 

6. Fill the missing values of `y` in `df_final` with zeros. Remove `STATUS` and `MONTHS_BALANCE` from `df_final`. (1 mark)

(10 marks)

In [6]:
# ---- provide your code here -----
# Question 2:
# 1. Change the values of STATUS in df according to the following mapping: {C, X, 0} -> 0 
# and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers.

status_mapping = {'C': 0, 'X': 0, '0': 0, '1': 1, '2': 1, '3': 1, '4': 1, '5': 1}
df['STATUS'] = df['STATUS'].map(status_mapping).astype(int)

In [7]:
# 2. Create a new numpy array called list_of_past_due 
# that includes the unique ID numbers of clients whose STATUS = 1 at any point during the last 12 months 
# (hint: count the current month as the first month).

recent_12_months = df[df['MONTHS_BALANCE'] >= -11]
list_of_past_due = recent_12_months[recent_12_months['STATUS'] == 1]['ID'].unique()

In [8]:
print(len(list_of_past_due))

1737


In [9]:
# 3. Create a new DataFrame called df_final that contains the rows of df for which the ID is in list_of_past_due, 
# keeping only one row for each ID (hint: keep the first duplicate row). 
# How many rows do you have in df_final? Answer using both print() function and in Markdown text. (hint: find out about isin() function in pandas.) 

df_final = df[df['ID'].isin(list_of_past_due)].drop_duplicates(subset=['ID'], keep='first')

In [10]:
print("Number of rows in df_final:", df_final.shape[0])

Number of rows in df_final: 1737


---
Number of rows in df_final: 1737

In [11]:
# 4. Add a new column y = 1 for all the rows in df_final. 

df_final['y'] = 1
#df_final.head()

In [12]:
# 5. Increase df_final to a total of 4,500 rows by adding rows from df with unique IDs which are not in list_of_past_due. 
# To do this start adding the rows from the beginning of df. (hint: learn what ~, i.e. tilde sign, does in pandas).

remain_id = df[~df['ID'].isin(list_of_past_due)]['ID'].unique()
add_rows = df[df['ID'].isin(remain_id)].drop_duplicates(subset=['ID'], keep='first')

df_final = pd.concat([df_final, add_rows]).head(4500)
#df_final.head(1800)

In [13]:
# 6. Fill the missing values of y in df_final with zeros. Remove STATUS and MONTHS_BALANCE from df_final.

df_final['y'] = df_final['y'].fillna(0).astype(int)
df_final = df_final.drop(columns=['STATUS', 'MONTHS_BALANCE'])

---- provide your written answer here ----


<hr style="width:25%;margin-left:0;"> 

**Question 3**. 
1. Delete `ID` column from `df_final` and reset its index. (1 marks)
2. Assuming that `NAME_EDUCATION_TYPE` is the only ordinal variable in `df_final`, which variables are numeric and which ones are nominal? Answer this question by copying and completing the following table (6 marks)

|Variable type|Numbers of features|Features' list|
| --- | --- | --- |
|Numeric:|||
|Ordinal:|1| NAME_EDUCATION_TYPE |
|Nominal:|||

3. Using appropriate functions find and comment on the missing values in `df_final` (3 marks)   

(10 marks)

In [14]:
# ---- provide your code here ----
# Question 3: 
# 1. Delete ID column from df_final and reset its index. 

df_final = df_final.drop(columns=['ID']).reset_index(drop=True)

In [25]:
# Assuming that NAME_EDUCATION_TYPE is the only ordinal variable in df_final, 
# which variables are numeric and which ones are nominal? 
# Answer this question by copying and completing the following table
df_final.head()

CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME  \
0           F            N               Y           NaN    157500.0   
1           M            Y               Y           NaN    360000.0   
2           F            N               Y           0.0    297000.0   
3           F            N               Y           0.0    297000.0   
4           F            N               Y           0.0    297000.0   

       NAME_INCOME_TYPE            NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  \
0               Working  Secondary / secondary special               Married   
1  Commercial associate                            NaN               Married   
2  Commercial associate                            NaN  Single / not married   
3  Commercial associate                            NaN  Single / not married   
4  Commercial associate                            NaN  Single / not married   

   NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  FLAG_WORK_PHONE  \
0  House / apartment      -10031          -1469           1                0   
1  House / apartment      -16670          -5364           1                0   
2   Rented apartment      -15519          -3234           1                0   
3   Rented apartment      -15519          -3234           1                0   
4   Rented apartment      -15519          -3234           1                0   

   FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  y  
0           1           0        Laborers                2  1  
1           1           0  Security staff                2  1  
2           0           0        Laborers                1  1  
3           0           0        Laborers                1  1  
4           0           0        Laborers                1  1

|Variable type|Numbers of features|Features' list|
| --- | --- | --- |
|Numeric:|10| CNT_CHILDREN, AMT_INCOME, DAYS_BIRTH, DAYS_EMPLOYED, FLAG_MOBIL, FLAG_WORK_PHONE, FLAG_PHONE, FLAG_EMAIL, CNT_FAM_MEMBERS, y |
|Ordinal:|1| NAME_EDUCATION_TYPE |
|Nominal:|7| CODE_GENDER, FLAG_OWN_CAR, FLAG_OWN_REALTY, NAME_INCOME_TYPE, NAME_FAMILY_STATUS, NAME_HOUSING_TYPE, OCCUPATION_TYPE |

In [29]:
# 3. Using appropriate functions find and comment on the missing values in df_final 

missing_values = df_final.isnull().sum()
print(missing_values)

CODE_GENDER               0
FLAG_OWN_CAR              0
FLAG_OWN_REALTY           0
CNT_CHILDREN             74
AMT_INCOME                0
NAME_INCOME_TYPE          0
NAME_EDUCATION_TYPE    1831
NAME_FAMILY_STATUS        0
NAME_HOUSING_TYPE         0
DAYS_BIRTH                0
DAYS_EMPLOYED             0
FLAG_MOBIL                0
FLAG_WORK_PHONE           0
FLAG_PHONE                0
FLAG_EMAIL                0
OCCUPATION_TYPE        1354
CNT_FAM_MEMBERS           0
y                         0
dtype: int64


---- provide your written answer here ----


---
---

### Task 2: Imputing missing values and dealing with categorical features (Total Marks: 30)



**Question 4.** 
1. Use an appropriate `pandas` function to impute missing values in `df_final` (15 marks)
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes

(Total: 15 marks)

In [15]:
# ---- provide your code here -----

<hr style="width:25%;margin-left:0;"> 

**Question 5**. Convert the values in `NAME_EDUCATION_TYPE` as follows
- Lower secondary -> 1
- Secondary / secondary special -> 2
- Incomplete higher -> 3
- Higher education -> 4


(Total: 5 marks)  

In [16]:
# ---- provide your code here -----

<hr style="width:25%;margin-left:0;"> 

**Question 6**. 

Add dummy variables to `df_final` for all of the nominal features which are currently stored as string (text). 
- Make sure to delete the original variables from the dataframe
- Drop the first column from each set of created dummy variable, i.e. for each feature



(Total: 10 marks)  

In [17]:
# ---- provide your code here -

---
---

### Task 3 Preparing X and y arrays (Total Marks: 10)

**Question 7**. 

1. Create a numpy array named `y` from the `y` column of `df_final` making sure that the values of the array `y` are stored as integers (3 marks)   
2. Create a numpy array named `X`  from all the remaining features in `df_final` (2 marks)   

(Total: 5 Marks)

In [18]:
# ---- provide your code here -----

<hr style="width:25%;margin-left:0;"> 

**Question 8**. 

1. Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 75% train and 25% test datasets (2.5 marks) 
    - Set random_state to 8 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
2. Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [19]:
# ---- provide your code here -----


---
---

### Task 4. Logistic Regression and Random Forest Classifiers and Accuracies (Total Marks: 30)


**Question 9**. 

1. Train a Logistic Regression Classifier on standardised data (5 marks)
    - Set `random_state` to 10 (don't change any other parameters)
    - Compute and print training and test dataset accuracies   
2. Train a Random Forest Classifier on standardised data (5 marks)
    - Set `random_state` to 10 (don't change any other parameters)
    - Compute and print training and test dataset accuracies

When printing accuracies round the values to three decimal places.      

(Total: 10 marks)  

In [20]:
# ---- provide your code here -----


**Question 10**. 

a) Comment and compare the training and test accuracies for each classifier computed in Question 9. What can we say about the extent of overfitting for each classifier? (5 marks)   
b) Comment and compare the accuracies across the two classifiers. Which classifier provides better forecasts? (5 marks)   
c) What can you say about the presence of nonlinearities in the dataset? (10 marks)   

(Total: 20 marks)  

---- provide your written answer here ----


---
---